### PR TITLE
Adding Not parser which negates a wrapped parser

### PR DESCRIPTION
**Pull request by ejsmith**
Starting a pull request with these changes for feedback.

### CODE REVIEW

File: Pidgin/Parser.Not.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        /// <returns>A parser which tries to apply the current parser, and applies <paramref name="parser"/> if the current parser fails without consuming any input.</returns>
+        public static Parser<TToken, TToken> Not<TToken>(Parser<TToken, TToken> parser) {
+            return new NegatedParser<TToken>(parser);
+        }
+        
+        private sealed class NegatedParser<TToken> : Parser<TToken, TToken>
+        {
+            private readonly Parser<TToken, TToken> _parser;
+
+            public NegatedParser(Parser<TToken, TToken> parser) : base(parser.Expected)
```
Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: The `base` call here says that the `Not` parser expects whatever the wrapped parser expects, which ain't right in this instance. Since we can't know statically what this parser does expect (only what it doesn't), I'd go with `ExpectedUtil<TToken>.Empty`.

File: Pidgin/Parser.Not.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+            internal sealed override Result<TToken, TToken> Parse(IParseState<TToken> state)
+            {
+                var startingPosition = state.SourcePos;
+                var result = _parser.Parse(state);
+                if (result.Success)
+                {
+                    return Result.Failure<TToken, TToken>(
+                            new ParseError<TToken>(
+                                new Maybe<TToken>(result.Value),
```
Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: We probably want to report the first token of the input as the unexpected thing, to match `startingPosition`. So I think you wanna call `state.Peek` (without advancing the stream) before `_parser` executes, and then report that token as the unexpected thing.

File: Pidgin/Parser.Not.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+namespace Pidgin
+{
+    public static partial class Parser
+    {
+        /// <summary>
+        /// Creates a parser which attempts to apply the specified parser and only succeeds if that parser fails.
+        /// </summary>
+        /// <param name="parser">The parser to apply if the current parser fails without consuming any input</param>
+        /// <returns>A parser which tries to apply the current parser, and applies <paramref name="parser"/> if the current parser fails without consuming any input.</returns>
+        public static Parser<TToken, TToken> Not<TToken>(Parser<TToken, TToken> parser) {
```
Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: I tried going that route, but the problem is that you end up needing a Map function because you don't know how to convert the current token for a successful parse to the return type.

File: Pidgin/Parser.Not.cs (hunk 1 above)
Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: Yeah, that makes sense. Changed.

File: Pidgin/Parser.Not.cs (hunk 2 above)
Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: Ok, changed.

File: Pidgin/Parser.Not.cs (hunk 3 above)
Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: Yeah, I think the return type would have to be `Unit`. See [my other comment](https://github.com/benjamin-hodgson/Pidgin/pull/4#discussion_r127315269)

File: Pidgin.Tests/ParserTestBase.cs [hunk 4]
```diff
@@ -19,5 +19,13 @@ public class ParserTestBase
...
             Assert.Equal(consumedInput, result.ConsumedInput);
         }
+
+        protected void AssertUnexpectedFailure<TToken, T>(Result<TToken, T> result, Maybe<TToken> unexpected, SourcePos errorPos, bool consumedInput)
+        {
+            Assert.False(result.Success);
+            Assert.Equal(unexpected, result.Error.Unexpected);
+            Assert.Equal(errorPos, result.Error.ErrorPos);
+            Assert.Equal(consumedInput, result.ConsumedInput);
+        }
```
Comment (ejsmith) [inline Pidgin.Tests/ParserTestBase.cs]: Do you really want to put everything in one method? It seems like you would either have a unexpected or an expected failure and they are different things.

File: Pidgin/Parser.Not.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                }
+
+                var token = state.Peek();
+                if (token.HasValue) {
+                    state.Advance();
+                    return Result.Success<TToken, TToken>(
+                            token.Value,
+                            true
+                        );
+                }
```
Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: Hmm... so in my use case, I am actually wanting it to consume the character that does not match the specific list of characters.

`Not(OneOf('a', 'b', 'c'))`

If not does not consume then all use cases would have to get expanded to:

`Not(OneOf('a, 'b', 'c')).Or(Any)`

Kind of makes it a bit less usable, but maybe that is where the derived helper functions come into play.

Hmm... I will have to dig into how the Unit type works.

Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: Having trouble getting anything to work when using Unit. Can you think of any real world use cases where you would want it to do what your suggested test is doing? It consumes "foo" and then bails because it doesn't match "d". Now you are sitting in the middle of the target string and it matches "bar". That seems very strange to me and seems like the only way it would be useful is to wrap it in a Try so it backtracks.

Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: I tried to make NoneOf work with Not returning a Parser<TToken, Unit> like this:

```csharp
public static Parser<char, char> NoneOf(params char[] chars)
            => Try(Not(OneOf((IEnumerable<char>)chars))).Or(Parser<char>.Token(c => true));
```

And I can't get it to work because the return types don't match.

Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: `NoneOf` would look like this:

```csharp
public static Parser<char, char> NoneOf(params char[] chars)
    => Not(OneOf(chars)).Then(Any);
```

`OneOf(chars)` doesn't consume input when it fails.

Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: > Can you think of any real world use cases where you would want it to do what your suggested test is doing?

It's the "principle of least surprise". I'd rather not make the decision to consume a character on behalf of the user, because I don't know what they're going to use `Not` for. If they _do_ want to start consuming characters after the wrapped parser fails, they can use `Then`; if they want to backtrack they can use `Try`.

File: Pidgin.Tests/ParserTestBase.cs (hunk 4 above)
Comment (benjamin-hodgson) [inline Pidgin.Tests/ParserTestBase.cs]: The vast majority of failures have both an `Unexpected` token and some `Expected` input. It means "here's what I was hoping for, but here's what I actually got". Perhaps the terminology could do with a little finessing, they do sound like "types of error" rather than properties of all errors.

Comment (ejsmith) [inline Pidgin.Tests/ParserTestBase.cs]: Are you wanting to add `Maybe<TToken> unexpected` as a parameter to `AssertFailure` then and then update all usages to pass the new parameter?

**Comment by ejsmith** [comment]
I just pushed a new commit with the changes you requested.

File: Pidgin.Tests/ParserTestBase.cs (hunk 4 above)
Comment (benjamin-hodgson) [inline Pidgin.Tests/ParserTestBase.cs]: Yeah, that's what I was thinking. (Don't do that work yourself, it'll be super boring 😜)

File: Pidgin/Parser.Not.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                var result = _parser.Parse(state);
+                if (result.Success)
+                {
+                    return Result.Failure<TToken, Unit>(
+                            new ParseError<TToken>(
+                                token,
+                                false,
+                                null,
+                                startingPosition,
+                                "unexpected token"
```
Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: One more (very minor) thing. Looking through the usages of this `message` parameter, it looks like it's primarily used for user-supplied messages. Almost all the other usages pass `null` here ([example](https://github.com/benjamin-hodgson/Pidgin/blob/master/Pidgin/Parser.Token.cs#L39)). Would you mind updating that? Then we can get this merged 👍

File: Pidgin/Parser.NoneOf.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+namespace Pidgin
+{
+    public static partial class Parser
+    {
+        /// <summary>
+        /// Creates a parser which parses and returns a character as long as it does not match one of the specified characters.
+        /// </summary>
+        /// <param name="chars">A sequence of characters that should not be matched</param>
+        /// <returns>A parser which parses and returns a character that does not match one of the specified characters</returns>
+        public static Parser<char, char> NoneOf(params char[] chars)
```
Comment (benjamin-hodgson) [inline Pidgin/Parser.NoneOf.cs]: I wonder if we can give this a more descriptive name? `AnyCharExcept`? Also, I would expect an implementation based on `Token` to be more efficient than one base on `OneOf` (but that's a discussion for another PR)

Comment (ejsmith) [inline Pidgin/Parser.NoneOf.cs]: Up to you, I was going with the name that you suggested earlier. I do think AnyCharExcept is a more descriptive name, but it might result in more method names if we want to have various overloads that take different kinds of parsers.

Comment (benjamin-hodgson) [inline Pidgin/Parser.NoneOf.cs]: Let's go with `AnyCharExcept`. It leaves `NoneOf` free for `Not(OneOf(...))`, which is how I would expect a parser named `NoneOf` to be implemented.

I think it would be a good idea to copy the input array, so as to safeguard against it getting mutated.

Comment (benjamin-hodgson) [inline Pidgin/Parser.NoneOf.cs]: Or, `AnyExcept` might be a good name too. What do you think?

Comment (ejsmith) [inline Pidgin/Parser.NoneOf.cs]: I think you are right that this is really unrelated to the Not changes and so I changed it up to not use `Not` and put the methods inside of the Parser.Char.cs file. I think AnyCharExcept seems to be the most descriptive.

File: Pidgin/Parser.Not.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                {
+                    return Result.Failure<TToken, Unit>(
+                            new ParseError<TToken>(
+                                token,
+                                false,
+                                null,
+                                startingPosition,
+                                null
+                            ),
+                            true
```
Comment (benjamin-hodgson) [inline Pidgin/Parser.Not.cs]: I think this should be `result.ConsumedInput`, not `true`. I guess there's a missing test here, which checks that `Not(Return(...))` doesn't consume input

File: Pidgin/Parser.NoneOf.cs (hunk 7 above)
Comment (benjamin-hodgson) [inline Pidgin/Parser.NoneOf.cs]: 👍🏻

File: Pidgin/Parser.Not.cs (hunk 8 above)
Comment (ejsmith) [inline Pidgin/Parser.Not.cs]: Updated

**Comment by ejsmith** [comment]
First! 😄  Thanks for taking this.

**Comment by benjamin-hodgson** [comment]
Really good work here. Thank you so much for the contribution and for your patience!

Do you have any feedback for me about the contributing process? Was it easy to get the code and tests set up on your machine? Was I a good reviewer?

**Comment by ejsmith** [comment]
I think your code is super clean and was easy to get started with. Only thing that gave me problems is the F# code being included in the project and I don't have F# installed. I wonder if maybe you should create a separate solution for the benchmarks.

And as far as the review process, you were super helpful.

I used vscode to make these changes. I may submit a pull request with the proper vscode configuration files.
